### ISSUE TITLE
How to use dice for a pipeline that performs one-hot-encoding when requesting prediction to tensorflow model

### ISSUE THREAD

**Issue by mike1130694** [interpretml/DiCE#43]
Hi,

I hope I can explain my current situation clear enough.

At the moment I've been trying to evaluate multiple model-agnostic methods to generate explanations on a Keras TensorFlow neural network model. Since each XAI method has different requirements and some need to have a numerical encoding to perform the perturbances on categorical data, 

The approach that I had implemented to be able to use multiple XAI methods is to have a "numerical encoded" dataset with 18 features (3 numerical and 15 categorical) where categories are encoded as integers. To feed this dataset to the keras NN model I used a column transformer pipeline that performs a ohe-hot-encoding of the 15 categorical variables and a min-max scaler for the numerical variables. Thus, when I call for a prediction i usually include the column transformer 

`nn.predict(ct.transform(x))`

I am trying to initialize the dice explainer however I haven't been able to figure out how I could connect this pipeline step to the model and provide it to dice. If I tried to load the ordinal encoded dataset and provide the NN model, I get an error during initialization stating

"Input 0 of layer dense_104 is incompatible with the layer: expected axis -1 of input shape to have value 66 but received input with shape [1, 65]"

My column transformer takes 18 features and transforms them in 66 columns. I will continue to debug, but I would appreciate if you could let me know if the current design of my pipeline could be integrated to DICE explainer.

Let me know if something is not clear enough and I will try to further explain.

Thanks a lot

**Comment by raam93** [comment]
Hi, 

One solution I can think of without changing DiCE's code is providing only the three numerical features to `continuous_features` parameter in `dice_ml.Data` and removing the column transformer component before prediction. If this is possible to do in your case, DiCE internally converts your 15 numerically-coded categorical features into `category` type. You can transform the categorical features in the output CFs to whatever labels you want. But check out the `predict_fn` in the [explainer](https://github.com/interpretml/DiCE/blob/master/dice_ml/explainer_interfaces/dice_tensorflow2.py#L97) and [model](https://github.com/interpretml/DiCE/blob/master/dice_ml/model_interfaces/keras_tensorflow_model.py) class to confirm if DiCE is getting prediction in the way you want (i.e., DiCE's gets predictions through `your_ann_model(input)`).

Of course, I have made several assumptions in the above solution about what you can change in your code. Let me know if this works else we can figure out a way by changing DiCE's internal code a little bit for your application.

By the way, I totally understand the challenges in comparing different explanation methods and good luck!

**Comment by amit-sharma** [comment]
Since v0.5, DiCE supports pipelines in the model. See this notebook, for example: https://github.com/interpretml/DiCE/blob/master/docs/source/notebooks/DiCE_getting_started.ipynb
Hope this resolves the issue.
